### DISCUSSION TITLE
Is there a way to propagate panic in actor? How do you write tests?

### DISCUSSION THREAD

**Discussion by nfejzic** [slawlor/ractor#289] [Q&A]
Hello! First of all, thank you for this great crate, appreciate it a lot. 

I have a question regarding testing. As far as I can see, the panics are caught and I don't see a way to figure out that an `Actor` panicked. When using actors in tests, this means that tests pass even though something panicked somewhere. Is there a way to make it so that panic gets propagated and makes the test fail?

**Comment by slawlor** [comment]
In order to determine if an actor panicked, there's essentially two flows.

## Flow 1: Actor panicked in `pre_start`

In this case, the panic will be caught and wrapped into an [`SpawnErr::StartupFailed`](https://github.com/slawlor/ractor/blob/main/ractor/src/errors.rs#L19) err. So the actor won't complete startup or register into the supervision flow

## Flow 2: Actor panics in either `post_start`, `post_stop`, `handle`, or `handle_supervisor_evt`

In this case, you would need the actor to have a supervisor. If there's a supervisor, the supervisor will have a `SupervisionEvent` sent to it and it'll be handled by the [`handle_supervisor_evt()`](https://github.com/slawlor/ractor/blob/main/ractor/src/actor/mod.rs#L312-L352) logic. If there is *no supervisor* then the panic is lost, and someone would need to wait on the `JoinHandle` to see the actor died and/or poll the actor's `ActorStatus` on the `ActorCell`|`ActorRef`. But you wouldn't be able to capture the panic message without a supervision flow.

**Comment by slawlor** [comment]
In a test-case, I'd wait for the actor to exit however if it actually failed, there isn't a clear way to capture the panic without a supervisor (and that's somewhat by design).

**Comment by nfejzic** [comment]
Thank you for your answer. Yeah, the fact that it's hard to crash the system is very nice to be honest. In testing it's a little inconvenient, but that's a good trade-off I would say. 

Another question/feature request would be, is there a way to stop the actor but don't stop processing messages already in the queue? The `stop_and_wait` stops processing immediately, and `handle.await` seems to just hang if there's no signal to stop the actor. What we do right now is we have a message available in tests only that sends a response. This makes it possible to have a sync point where we wait for the message to process. 

A functionality that signals to the actor to stop receiving new messages, finish processing the ones in the queue and then stop would be great!

**Comment by slawlor** [comment]
There is, it's called drain or drain_and_wait. You might just need to upgrade your version if you don't have it already

**Comment by nfejzic** [comment]
> There is, it's called drain or drain_and_wait. You might just need to upgrade your version if you don't have it already

Ah alright, awesome! I will update and try it out, thank you!
